benchmark/utils/stencil_matrix.hpp

```diff
@@ -0,0 +1,577 @@
+ *                           deviate from this to accommodate the uniform size
+ *                           of the boxes.
+ * @param restricted  If true, a 7-pt stencil is used, else a 27-pt stencil.
+ *
+ * @return  matrix data of a box using either 7-pt or 27-pt stencil.
+ */
+template <typename ValueType, typename IndexType>
+gko::matrix_data<ValueType, IndexType> generate_3d_stencil_box(
+    std::array<int, 3> dims, std::array<int, 3> positions,
+    const gko::size_type target_local_size, bool restricted)
+{
+    auto num_boxes =
+        std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<>{});
+
+    const auto discretization_points =
+        static_cast<IndexType>(closest_nth_root(target_local_size, 3));
+    const auto local_size = static_cast<gko::size_type>(
+        discretization_points * discretization_points * discretization_points);
+    const auto global_size = local_size * num_boxes;
+    auto A_data = gko::matrix_data<ValueType, IndexType>(
+        gko::dim<2>{static_cast<gko::size_type>(global_size),
+                    static_cast<gko::size_type>(global_size)});
+
+    auto global_offset = [&](const int position_x, const int position_y,
+                             const int position_z) {
+        return position_x * static_cast<int>(local_size) +
+               position_y * static_cast<int>(local_size) * dims[0] +
+               position_z * static_cast<int>(local_size) * dims[0] * dims[1];
+    };
+
+    auto target_position = [&](const IndexType i, const int position) {
+        return is_in_box(i, discretization_points)
+                   ? position
+                   : (i < 0 ? position - 1 : position + 1);
+    };
+
+    auto target_local_idx = [&](const IndexType i) {
+        return is_in_box(i, discretization_points)
+                   ? i
+                   : (i < 0 ? discretization_points + i
+                            : discretization_points - i);
+    };
+
+    auto flat_idx = [&](const IndexType ix, const IndexType iy,
+                        const IndexType iz) {
+        auto tpx = target_position(ix, positions[0]);
+        auto tpy = target_position(iy, positions[1]);
+        auto tpz = target_position(iz, positions[2]);
+        if (is_in_box(tpx, dims[0]) && is_in_box(tpy, dims[1]) &&
+            is_in_box(tpz, dims[2])) {
+            return global_offset(tpx, tpy, tpz) + target_local_idx(ix) +
+                   target_local_idx(iy) * discretization_points +
+                   target_local_idx(iz) * discretization_points *
+                       discretization_points;
+        } else {
+            return static_cast<IndexType>(-1);
+        }
+    };
+
+    auto is_valid_neighbor = [&](const IndexType d_i, const IndexType d_j,
+                                 const IndexType d_k) {
+        return !restricted ||
+               ((d_i == 0 && d_j == 0) || (d_i == 0 && d_k == 0) ||
+                (d_j == 0 && d_k == 0));
```

Comment:
we can compute the Manhattan distance directly
```suggestion
               ((d_i == 0) + (d_j == 0) + (d_k == 0) == 1);
```

Reply:
oops yes, I was thinking of != 0. If you wanted to code-golf this, you could also do 
```suggestion
               (!d_i+ !d_j + !d_k == 2);
```
but that's hard to understand at a glance :smile: